apps/prairielearn/src/models/lti-user.ts

```diff
@@ -0,0 +1,108 @@
+import { z } from 'zod';
+
+import { callRow } from '@prairielearn/postgres';
+
+import { insertAuditEvent } from './audit-event.js';
+import { ensureEnrollment } from './enrollment.js';
+import { selectInstitutionByShortName } from './institution.js';
+import { insertUserLti, selectUserByUidAndInstitution, updateUserName } from './user.js';
+
+export interface LtiUserResult {
+  userId: string;
+  hasAccess: boolean;
+}
+
+export async function selectOrInsertAndEnrollLtiUser({
+  uid,
+  name,
+  lti_course_instance_id,
+  lti_user_id,
+  lti_context_id,
+  req_date,
+}: {
+  uid: string;
+  name: string | null;
+  lti_course_instance_id: string;
+  lti_user_id: string;
+  lti_context_id: string;
+  req_date: Date;
+}): Promise<LtiUserResult> {
+  const { id: institution_id } = await selectInstitutionByShortName({ short_name: 'LTI' });
+
+  let user = await selectUserByUidAndInstitution({
+    uid,
+    institution_id,
+  });
+
+  if (!user) {
+    user = await insertUserLti({
+      uid,
+      name,
+      lti_course_instance_id,
+      lti_user_id,
+      lti_context_id,
+      institution_id,
+    });
+
+    await insertAuditEvent({
+      action: 'insert',
+      table_name: 'users',
+      row_id: user.user_id,
+      new_row: user,
+      institution_id,
+      course_instance_id: lti_course_instance_id,
+      subject_user_id: user.user_id,
+      agent_authn_user_id: null,
+    });
+  }
+
+  // Update user name if needed
+  if (name !== null && name !== user.name) {
+    const oldUser = user;
+    user = await updateUserName({
+      user_id: user.user_id,
+      name,
+    });
+
+    await insertAuditEvent({
+      action: 'update',
+      action_detail: 'name',
```

Comment:
PT has a `column_name` field for this purpose, but I see we don't. Did you consider that when implementing the schema?

I don't want to pretend that PT got everything exactly right. But IMO it _did_ get a lot right, and it's going to be very confusing to all of us if we have subtly different conventions/setups between the projects. I'd like to avoid that if possible; that's why I keep bringing up PT.

cc @mwest1066 if you have thoughts on this